### DISCUSSION TITLE
Unable to add child item in M2A reference with Graphql mutation query

### DISCUSSION THREAD

**Discussion by SPJames** [directus/directus#5178] [Archive]
1. Problem
When following the [API Reference docs](https://docs.directus.io/reference/api/introduction/#one-to-many-many-to-many-many-to-any) I would expect when using a mutation query I'm able to add a child item to a M2A reference. However this causes an error.
I have an item called Pages, which contains a M2A reference to multiple types of layout items. Using the following graphql mutation query I would expect it would create a new M2A reference: 

```graphql
mutation {
   update_pages_item(id: <PAGE_ID>, data: {
     layouts: [<NEW_LAYOUT_ID>]
   }) {
     id
   }
}
``` 

This query throws the following error: 

```json
{
   "message":"GraphQL validation error.",
   "extensions":{
      "code":"GRAPHQL_VALIDATION_EXCEPTION",
      "graphqlErrors":[
         {
            "message":"Expected value of type ""update_pages_layouts_input"", found 20.",
            "locations":[
               {
                  "line":4,
                  "column":21
               }
            ]
         }
      ]
   }
}
```

2. Reproduce
In order to reproduce this first create an item which will get a M2A reference to at least 2 items.
Create the 2 child items. Add them to the M2A reference.
Using an external Graphql tool use the mutation query above to insert a newly created child item into the M2A reference.

3. Versions:
Version 9.0.0-rc59 (but this problem has been there from the version 9.0.0-rc54 when the Graphql 2.0 was introduced)
Postgres 12
Running in Docker (directus and database are different containers)

4. Extra information
I'm creating a webapp using Angular, however I don't think this has any impact.

**Comment by rijkvanzanten** [comment]
I believe that `layouts` in this example is a o2m to the junction table (presumably named something like `pages_layouts`. Have you tried:

```graphql
mutation {
	update_pages_layout(data: {
		layouts: [
			{
				layouts_id: <NEW_LAYOUT>
			}
		]
	}) {
		id
	}
}
```

?

**Reply by SPJames** [reply]
I got the SDL from the server to check how the input should look like. Cause you are right that is a junction table. However I'm still unable to create it. Below are the graphql sdl
```graphql
input update_pages_input {
  id: Int
  status: String
  user_created: update_directus_users_input
  date_created: Date
  user_updated: update_directus_users_input
  date_updated: Date
  title: String
  body: String
  slug: String
  layouts: [update_pages_layouts_input]
}

input update_pages_layouts_input {  
  id: Int  
  pages_id: update_pages_input  
  collection: String  
  sort: Int  
  item: String  
}
```

Seeing this I would suspect that `item` would reference the layout, however it's a String. When looking at the pages_layout sdl itself I see this: 
```graphql
type pages_layouts {
  id: Int
  pages_id(filter: pages_filter, sort: [String], limit: Int, offset: Int, page: Int, search: String): pages
  collection: String
  sort: Int
  item: pages_layouts_item_union
}
union pages_layouts_item_union = grid_layout | carousel_layout | column_layout | single_layout
```
Which seems to be correct, as in the `item` references one of the possible `layouts`.
Is this a bug or am I missing something else?

**Reply by rijkvanzanten** [reply]
GraphQL doesn't support union inputs, so for the nested creates, you can't nest under `item`. `item: String` is correct in this case, as it's the "raw" foreign key to the related any item 👍🏻 

Groetjes!

**Reply by krazyjakee** [reply]
So is the conclusion that there is no way to create references between items using graphql?
I can't get any m2a or m2m example to work without errors.

Using REST post I can get it to go through but it sets the reference to an integer ID which shows as empty on the directus edit form.

**Reply by rijkvanzanten** [reply]
@krazyjakee This discussion is over 8 months old. A lot has changed since the pre-release versions. 

> So is the conclusion that there is no way to create references between items using graphql?

No, the information on https://docs.directus.io/reference/introduction/#relational-data is accurate, which includes examples of nested mutations on GraphQL. 

> Using REST post I can get it to go through but it sets the reference to an integer ID which shows as empty on the directus edit form.

This sounds like a (potential) bug. Mind [opening a new issue](https://github.com/directus/directus/issues/new) if it wasn't already reported?
